spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/inbound/AbstractHazelcastMessageProducer.java

```diff
@@ -0,0 +1,200 @@
+				getDistributedObject(), this.getCacheEventTypeSet());
+	}
+
+	protected Set<String> getCacheEventTypeSet() {
+		return StringUtils.commaDelimitedListToSet(getCacheEventTypes());
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		HazelcastIntegrationUtils.shutdownAllHazelcastInstances();
+	}
+
+	protected DistributedObject getDistributedObject() {
+		return distributedObject;
+	}
+
+	protected String getCacheEventTypes() {
+		return cacheEventTypes;
+	}
+
+	public void setCacheEventTypes(String cacheEventTypes) {
+		this.cacheEventTypes = cacheEventTypes;
+	}
+
+	protected CacheListeningPolicyType getCacheListeningPolicy() {
+		return cacheListeningPolicy;
+	}
+
+	public void setCacheListeningPolicy(CacheListeningPolicyType cacheListeningPolicy) {
+		this.cacheListeningPolicy = cacheListeningPolicy;
+	}
+
+	protected String getHazelcastRegisteredEventListenerId() {
+		return hazelcastRegisteredEventListenerId;
+	}
+
+	public void setHazelcastRegisteredEventListenerId(String hazelcastRegisteredEventListenerId) {
+		this.hazelcastRegisteredEventListenerId = hazelcastRegisteredEventListenerId;
+	}
+
+	protected abstract class AbstractHazelcastEventListener {
+
+		protected abstract void processEvent(EventObject event);
+
+		protected void sendMessage(EventObject event, InetSocketAddress socketAddress, CacheListeningPolicyType cacheListeningPolicyType) {
+			if (CacheListeningPolicyType.ALL == cacheListeningPolicyType || isEventAcceptable(socketAddress)) {
+				AbstractHazelcastMessageProducer.this.sendMessage(MessageBuilder.withPayload(event).build());
+			}
+		}
+
+		private boolean isEventAcceptable(InetSocketAddress socketAddress) {
+			Set<SocketAddress> localSocketAddressesSet = getLocalSocketAddresses();
+			if ((!localSocketAddressesSet.isEmpty())
+					&& (localSocketAddressesSet.contains(socketAddress) ||
+							this.isEventComingFromNonRegisteredHazelcastInstance(localSocketAddressesSet, socketAddress))) {
+				return true;
+			}
+
+			return false;
+		}
+
+		private Set<SocketAddress> getLocalSocketAddresses() {
+			Set<SocketAddress> localSocketAddressesSet = new HashSet<>();
+			for (HazelcastInstance hazelcastInstance : Hazelcast.getAllHazelcastInstances()) {
+				localSocketAddressesSet.add(hazelcastInstance.getLocalEndpoint().getSocketAddress());
+			}
+
+			return localSocketAddressesSet;
+		}
+
+		private boolean isEventComingFromNonRegisteredHazelcastInstance(Set<SocketAddress> localSocketAddressesSet, InetSocketAddress socketAddressOfEvent) {
+			HazelcastInstance hazelcastInstance = Hazelcast.getAllHazelcastInstances().iterator().next();
+			MultiMap<SocketAddress, SocketAddress> configMultiMap = hazelcastInstance
+					.getMultiMap(HazelcastLocalInstanceHandler.HZ_INTERNAL_CONFIGURATION_MULTI_MAP);
+			if (configMultiMap.size() > 0
+					&& !configMultiMap.values().contains(socketAddressOfEvent)
+					&& localSocketAddressesSet.contains(configMultiMap.keySet().iterator().next())) {
```

Comment:
I guess we'll come back to the `round-robin` discussion.
And, please, be sure that your code line aren't more than `120` symbols.
Plus I'd prefer to rename that `HZ_INTERNAL_CONFIGURATION_MULTI_MAP` to more appropriate value, since it is some cluster-wide object.